### PR TITLE
Allow empty partition keys in views

### LINKED ISSUE #3262: Database fails to start due to sstable confusion
*Installation details*
Scylla version 2.1.rc3-0.20180207.b98d5b30d
Cluster size: 1
OS (RHEL/CentOS/Ubuntu/AWS AMI): AWS i3.xlarge

We had some issues with materialized views a few weeks ago and I ended up needing to move aside the views and indexes directories to save the cluster.  Scylla recreate those directories and we were able to subsequently recreated the views and indexes from our saved schema and repopulated the data by updating the base tables.  However, we later started getting periodic error messages in the journalctl log about missing sstables:

```
Mar 07 06:29:38 scylladb scylla[24302]:  [shard 0] database - failed to write sstable /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-64-Data.db: std::runtime_error (first key of summary of /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-64-Data.db is empty)
Mar 07 06:29:48 scylladb scylla[24302]:  [shard 0] database - failed to write sstable /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-68-Data.db: std::runtime_error (first key of summary of /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-68-Data.db is empty)
Mar 07 06:29:58 scylladb scylla[24302]:  [shard 0] database - failed to write sstable /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-72-Data.db: std::runtime_error (first key of summary of /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-72-Data.db is empty)
Mar 07 06:30:08 scylladb scylla[24302]:  [shard 0] database - failed to write sstable /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-76-Data.db: std::runtime_error (first key of summary of /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-76-Data.db is empty)
Mar 07 06:30:18 scylladb scylla[24302]:  [shard 0] database - failed to write sstable /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-80-Data.db: std::runtime_error (first key of summary of /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-80-Data.db is empty)
Mar 07 06:30:28 scylladb scylla[24302]:  [shard 0] database - failed to write sstable /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-84-Data.db: std::runtime_error (first key of summary of /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-84-Data.db is empty)
Mar 07 06:30:38 scylladb scylla[24302]:  [shard 0] database - failed to write sstable /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-88-Data.db: std::runtime_error (first key of summary of /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-88-Data.db is empty)
Mar 07 06:30:48 scylladb scylla[24302]:  [shard 0] database - failed to write sstable /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-92-Data.db: std::runtime_error (first key of summary of /var/lib/scylla/data/veraminetest/tracked_urls_usersid_idx_index-0324aac01dde11e89244000000000002/veraminetest-tracked_urls_usersid_idx_index-ka-92-Data.db is empty)
```

In effort to resolve this issue, I tried nodetool drain.  The drain operation never finished, instead getting stuck in a loop failing to write to those sstables.  I also tried systemctl restart scylla-server and that also never completes.  It begins to replay the pending Commitlog files and eventually gets into a state where it is attempting to access missing sstables every 10 seconds and finally restarts the loop all over again.

I don't need the data in these tables, indexes, or views.  Is there a way I can just wipe out references to the "stuck" indexes?  Or, any other suggestions to best resolve this state?  I attached the most recent couple days of journalctl output, in case that helps.  Thanks!

[journal.txt](https://github.com/scylladb/scylla/files/1788071/journal.txt)

### LINKED ISSUE #9352: Different token calculation functions for "empty" partition keys cause sstable reader to use index incorrectly
I'm not sure if this causes an actual bug in production code because I don't know if "empty" partition keys can occur naturally (couldn't create them from CQL), but posting this just in case.

I have a schema with `blob` partition key type:
```
CREATE TABLE ... (
        pk0 blob,
        ....,
        PRIMARY KEY (pk0))
```

I have a value of this key in `dht::decorated_key` format:
```
dht::decorated_key k{...};
std::cout << k << std::endl;
```
it prints as follows:
```
{key: pk{0000}, token:0}
```

When I `explode` it, I get a vector of size `1` containing an empty `bytes`:
```
auto vec = k.explode(s);
std::cout << vec.size() << " " << vec << " " << vec[0].size() << std::endl;
```
result:
```
1 {} 0
```

So apparently this is a key with a single column whose value is an empty blob; the key is represented as `0000` in serialized form. I'm guessing this `0000` denotes the blob's size. I got this key from a mutation that was generated by `generate_random_mutations` in test/lib/random_schema.hh.

In a test, I created an sstable using this and some other mutations. This sstable's partition key index contains a single page, and among the entries in this page is an entry with this key:
```
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable - index 0x60100001e000 bound 0x60100001e070: page:
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable -   {key: pk{0010fffe81a2d0a483b760eba3bfa395e5d8}, token:-7666695272245969689} -> 0
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable -   {key: pk{0011ac795b361b3218b7a479b98132c067ef22}, token:-7047076813233896397} -> 4016
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable -   {key: pk{00075c536477ad7842}, token:-6422108399605007642} -> 10431
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable -   {key: pk{001ebd5ee00751bbec821cee44e7e6055ffa6d0ea50b3342e736511e85eddaca}, token:-5731315764860802051} -> 15622
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable -   {key: pk{00175fc396083983d96b8866cb88957439a3dfc82d1b7b71fb}, token:-5080536638769575688} -> 21953
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable -   {key: pk{001b54b4dd6b7afd961ce544da04fccb7db04e94263d51f0bea843619c}, token:-2826028800328955929} -> 34517
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable -   {key: pk{0016581a5b0b58ec1f25e2f0cc648e76c5e2fa4a4f07fa11}, token:-2358308865434803126} -> 39082
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable -   {key: pk{001657ca9f021164ef6584a08ab5bf8d97f03187a5f4f426}, token:-1618278504687846796} -> 45934
TRACE 2021-09-17 12:17:49,781 [shard 0] sstable -   {key: pk{0000}, token:0} -> 50158
```

When looking for a partition, the index reader algorithm uses a different class (not `dht::decorated_key`) to interpret the keys: `sstables::decorated_key_view`. While `dht::decorated_key` (which is a key, token pair) uses `partition_key` to store the key internally, `sstables::decorated_key_view` (which is also a key, token pair) uses `sstables::key_view`. The index calculates the tokens on the fly:
```
class index_entry {
private:
    managed_bytes _key;
    mutable std::optional<dht::token> _token;
    ....
    key_view get_key() const {
        return key_view{_key};
    }

    // May allocate so must be called under allocating_section.
    decorated_key_view get_decorated_key(const schema& s) const {
        if (!_token) {
            _token.emplace(s.get_partitioner().get_token(get_key()));
        }
        return decorated_key_view(*_token, get_key());
    }
```

There are two `i_partitioner::get_token` functions, one for `partition_key`s and one for `sstables:key_view`s:
```
    virtual token get_token(const schema& s, partition_key_view key) const = 0;
    virtual token get_token(const sstables::key_view& key) const = 0;
```

The first version obtains token `0` for our key:
```
token murmur3_partitioner::get_token(const schema& s, partition_key_view key) const {
    std::array<uint64_t, 2> hash;
    auto&& legacy = key.legacy_form(s);
    utils::murmur_hash::hash3_x64_128(legacy.begin(), legacy.size(), 0, hash);
    return get_token(hash[0]);
}
```

but the second obtains `minimum_token()`:
```
token murmur3_partitioner::get_token(const sstables::key_view& key) const {
    return key.with_linearized([&] (bytes_view v) {
        return get_token(v);
    });
}

token murmur3_partitioner::get_token(bytes_view key) const {
    if (key.empty()) {
        return minimum_token();
    }
    std::array<uint64_t, 2> hash;
    utils::murmur_hash::hash3_x64_128(key, 0, hash);
    return get_token(hash[0]);
}
```
because it enters the `key.empty()` case.

The index uses `index_entry::get_decorated_key`, hence the second version of `get_token` when performing a binary search when looking for a partition in the index page (sstables/index_reader.hh):
```
        return advance_to_page(bound, summary_idx).then([this, &bound, pos, summary_idx] {
            auto i = _alloc_section(_region, [&] {
                return std::lower_bound(std::begin(entries) + bound.current_index_idx, std::end(entries), pos,
                    index_comparator(*_sstable->_schema));
            });
```

with the page printed above, this is a binary search performed on a non-monotonic vector. This is because the last entry is considered to have the token `minimum_token()`, which compares before all other entries, but it is placed at the end!

This may cause the sstable reader to consider a partition missing even when it's there.

---

Two questions:
- is this even a "real" problem? Can such keys occur in production?
- how to fix this? Which token calculation function is wrong? Should we perform additional validation to detect this stuff?

Ad. second bullet, I tried to catch the `minimum_token()` at the sstable writer level (in `mutation_fragment_stream_validator`). However it doesn't work, because at that point we're using the `dht::decorated_key` format, so the token of this key here is calculated as `0`.

cc @tgrabiec @denesb

### LINKED ISSUE #9364: Empty strings are not indexed properly leading to false negatives when querying for empty values
```
CREATE TABLE cf ( name text PRIMARY KEY, val1 text, val2 text ); 
INSERT INTO cf (name, val1, val2) VALUES ('dmeyer', 'testval', '');
CREATE INDEX cf_val2_idx ON cf (val2);
```

```
cqlsh:ks> select * from cf;

 name    | val1    | val2
---------+---------+------
  dmeyer | testval |     


(1 row)

cqlsh:ks> select * from cf where val2 = '' allow filtering;

 name | val1 | val2
------+------+------

(0 rows)
```

After dropping the index the query works:

```
cqlsh:ks> drop index cf_val2_idx ;
```

```
cqlsh:ks> select * from cf where val2 = '' allow filtering;

 name   | val1    | val2
--------+---------+------
 dmeyer | testval |     

```

Refs #9352

### LINKED ISSUE #9375: Empty strings values are missing as a materialized view partition key
This issue is the underlying cause of #9364 (which is the same problem with secondary-indexes) and also closely related to the overall issue of empty-strings as partition key #9352.

For various reasons discussed in #9352, Scylla and Cassandra do not allow an empty string as a partition key in base tables (see also the test `test_null.py`::`test_empty_string_key`). 
However, an empty string is a _valid value_ for a non-key string column. So if we have a materialized view with a string column becoming the view's partition key - an empty string value may end up being the view row's partition key! The "IS NOT NULL" clause in the view's declaration does not eliminate this row (an empty string is _not_ considered NULL), so this row should appear in the view - complete with an empty-string partition key.
It turns out that this does happen in Cassandra, but not in Scylla. 

This issue was first noticed by @psarna in https://github.com/scylladb/scylla/issues/9364#issuecomment-924674990 and can be reproduced with the following cql-pytest:
```python
def test_mv_empty_string_partition_key(cql, test_keyspace):
    schema = 'p int, v text, primary key (p)'
    with new_test_table(cql, test_keyspace, schema) as table:
        with new_materialized_view(cql, table, '*', 'v, p', 'v is not null and p is not null') as mv:
            cql.execute(f"INSERT INTO {table} (p,v) VALUES (123, '')")
            # The view row with the empty partition key exists:
            # This is where Scylla currently fails
            assert list(cql.execute(f"SELECT * FROM {mv}")) == [('', 123)]
            # However, it is still impossible to select just this row,
            # because Cassandra forbids an empty partition key on select
            with pytest.raises(InvalidRequest, match='Key may not be empty'):
                cql.execute(f"SELECT * FROM {mv} WHERE v=''")
```

Note how curiously while Cassandra does store the view row with the empty-string partition key, and it is visible in a full-table scan, one cannot be read just this row directly with `SELECT * FROM mv WHERE v=''` because `WHERE v=''` for the partition key (of the view) `v` is considered an error. So arguably, this Cassandra behavior is **not useful** (the data is stored but cannot be efficiently read), and perhaps even not intentional (I can't find any documentation or discussion for this case).

So another option open to use is to decide (but we need to carefully consider this!) that Cassandra's behavior here is not useful, and decide (and document!) that Scylla's existing behavior is the right one. We can claim that the "IS NOT NULL" specification includes, as a **special case**, the requirement that a would-be partition key is also not an empty string. However, this would look rather arbitrary and strange.

### LINKED ISSUE #10178: sstables::key_view::tri_compare() incorrect when both keys are empty
The method `sstables::key_view::tri_compare(const schema& s, partition_key_view other)` fails to return `equal` as expected when given two **empty** keys, and instead returns `less`.

The following unit test reproduces this bug:
```c++
BOOST_AUTO_TEST_CASE(test_empty_key_view_comparison) {
    auto s_ptr = schema_builder("", "")
            .with_column("p", bytes_type, column_kind::partition_key)
            .build();
    const schema& s = *s_ptr;

    sstables::key empty_sstable_key = sstables::key::from_deeply_exploded(s, {data_value(bytes(""))});
    sstables::key_view empty_sstable_key_view = empty_sstable_key;
    partition_key empty_partition_key = partition_key::from_deeply_exploded(s, {data_value(bytes(""))});
    partition_key_view empty_partition_key_view = empty_partition_key;

    BOOST_CHECK_EQUAL(std::strong_ordering::equal,
        empty_sstable_key_view.tri_compare(s, empty_partition_key_view));
```

This bug will cause user-visible bugs after fixing https://github.com/scylladb/scylla/issues/9375 which allows empty-string partition keys in materialized views. When looking up a partition with an empty-string key, the index code tries to compare the empty key looked up to the empty key stored in the index file, and because the result isn't `equal`, the partition can't be found.

Interestingly, while we return incorrect results when comparing two empty keys, we do return the correct results when just one of the keys is empty (the correct result is that the empty key should be less than the non-empty key).

### PR DESCRIPTION
**Pull request by nyh**
Cassandra generally does not allow empty strings as partition keys (note, by the way, that empty strings are allowed as clustering keys, as well as in individual components of a compound partition key).
However, Cassandra does allow empty strings in _regular_ columns - and those regular columns can be indexed by a secondary index, or become an empty partition-key column in a materialized view. As noted in issues #9375 and #9364 and verified in a few xfailing cql-pytest tests, Scylla didn't allow these cases - and this patch series fixes that.

Before the last patch in this series finally enables empty-string partition keys in materialized views, we first need to solve a couple of bugs in our code related to handling empty partition keys:

The first patch fixes issue #10178 - a bug in `key_view::tri_compare()`  where comparing two empty keys returned a random result instead of "equal".

The second patch fixes issue #9352: our tokenizer has an inconsistency where for an empty string key, two variants of the same function return different results:

1. One variant `murmur3_partitioner::get_token(bytes_view key)` returned `minimum_token()` for the empty string. 
2. Another variant `murmur3_partitioner::get_token(const schema& s, partition_key_view key)` did not have this special case, and called the normal hash-function calculation on the empty string (the resulting token is 0).

Variant 2 was an unintentional bug, because Cassandra always does what variant does 1. So the "obvious" fix here would be to fix variant 2 to do what variant 1 does. Nevertheless, we decided to do the opposite: Change variant 1 to match variant 2. The reasoning is as follows:

The `minimum_token()` is `token{token::kind::before_all_keys, 0 }` - it's not a real token. Since we intend in this patch allow real data to exist with the empty key, we need this real data to have a real token. For example, this token needs to be located on the token ring (so the empty-key partition will have replicas) and also belong to one of the shards, and it's not clear that `minimum_token()` will be handled correctly in this context.

After changing the token of the empty string to 0, we note that some places in the code assume that `dht::decorated_key(dh
t::minimum_token(), partition_key::make_empty())` is a legal decorated key. However, as far as I can tell, none of these places actually assume that the partition-key part (the `make_empty()`) really matches the token - this decorated key is only used to start an iteration (ignoring this key itself) or to indicate a non-existent key (in modern code `std::optional` should be used for that).

While normally changing the token of a key is a big faux-pas, which can result in old data no longer being readable, in this case this change is safe because:

1. Scylla previously disallowed empty partition keys (in both base tables and views), so we cannot have had such a partition key saved in any sstable.
3. Cassandra does allow empty partition keys in _views_ and _secondary indexes_, but we do not support migrating sstables of those into Scylla - users are expected to only migrate the base table and then re-create the view or index. So however Cassandra writes those empty-key partitions, we don't care.

The third patch finally fixes the materialized views implementation to not drop view rows with an empty-string partition key (#9375). This means we basically revert commit ec8960df45fa7e64cb7548ffffb9e99fec06700c - which fixed #3262 by disallowing empty partition keys in views, whereas this patch fixes the same problem by handling the empty partition keys correctly.

The fix for the secondary index bug (#9364) comes "for free" because it is based on materialized views.

We already had xfailing test cases for empty strings in materialized views and indexes, and after this series they begin to pass so the "xfail" mark is removed. The series also adds additional test cases that validate additional corner cases discovered during the debugging.

Fixes #9352
Fixes #9364
Fixes #9375
Fixes #10178

### CODE REVIEW

**Comment by nyh** [comment]
This is a new pull-request doing the same thing done by pull-request #9528, just in a different way - using the token 0 for the empty key - not minimum_token, as @avikivity suggested in his review of #9528.

The pull-request is still in draft form because although I outlined in the first message what I want to do, I haven't actually done this yet, and I just started with commits copied from #9528.

**Comment by avikivity** [comment]
Is 0 hardcoded? Or do we always the hash function, and 0 happens to be the result?

**Comment by nyh** [comment]
> Is 0 hardcoded? Or do we always the hash function, and 0 happens to be the result?

It's not hard-coded - it just calls the murmur3 hash function as usual on the string - I removed the if() for checking the empty string - and 0 is the current result of the hash function on an empty string. This 0 result can never change, because the murmur3 function can never change without breaking all existing data.

File: compound_compat.hh [hunk 1]
```diff
@@ -71,8 +71,9 @@ public:
 
         iterator(const legacy_compound_view& v)
             : _singular(v._type.is_singular())
-            , _offset(_singular ? 0 : -2)
-            , _i(v._type.begin(v._packed))
+            , _offset(_singular && v._type.begin(v._packed)->size() ? 0 : -2)
```
Comment (avikivity) [inline compound_compat.hh:74]: This smells wrong. Why should we have -2 for an empty key? It yields the correct result but is not natural to have -2 offset (for the size bytes) when they won't be emitted.

I think it's cleaner to keep this 0, but adjust the end_tag constructor to set _offset to 0 for singular keys.

File: test/boost/view_schema_pkey_test.cc [hunk 2]
```diff
@@ -715,6 +715,12 @@ SEASTAR_TEST_CASE(test_base_non_pk_columns_in_view_partition_key_are_non_emtpy)
                     });
         }
 
+        // The following if'ed-out tests verified behavior that we
+        // no longer believe to be correct, and also turns out not to be
+        // compatible with Cassandra (see issue #9375): these tests checked
+        // that if a view row should get an empty string as a partition key,
+        // this row should not be generated. But this is not the case.
+#if 0
```
Comment (avikivity) [inline test/boost/view_schema_pkey_test.cc:723]: delete?

File: compound_compat.hh (hunk 1 above)
Comment (psarna) [reply compound_compat.hh:74]: It's preexisting (see lines 92-94 in this file), and moreover, it's specified like that in CQL. https://github.com/apache/cassandra/blob/ac34f28e31bbc294db80e2b5d06f8904a7f41a46/doc/native_protocol_v4.spec#L229-L233 Weird, but we need to remain compatible nonetheless

Comment (avikivity) [reply compound_compat.hh:74]: No, this is a different -2. This -2 leaves 2 bytes space for the size of the component. See operator*().

Comment (psarna) [reply compound_compat.hh:74]: Oh, right, this isn't CQL. Still, operator* already had this branch - so this is a hack, but still preexisting, for the sake of some other corner case? In that case we should probably only fix that in a follow-up

Comment (psarna) [reply compound_compat.hh:74]: For reference, Tomek was kind enough in 2015 to leave a graphical comment about these special values:
https://github.com/scylladb/scylla/blob/f2060ac03bf008b40bc8d0ece005b79745ca593b/compound_compat.hh#L62-L67

Comment (avikivity) [reply compound_compat.hh:74]: The -2 is fine for compound keys, but not singular keys. See my suggestion above (clarified here):

I think it's cleaner to keep this 0 _for singular keys_, but adjust the end_tag constructor to set _offset to 0 for singular keys.

Comment (nyh) [reply compound_compat.hh:74]: @avikivity right, that's also an option. I actually considered it but both seemed equally ugly :-( I went with my solution because it meant adding ugly conditionals only to the begin tag (which already had one such ugly conditional), and not the the end tag. But I'll prepare a new version where I uglify both - to reach a more balanced ugliness ;-)

File: test/boost/view_schema_pkey_test.cc (hunk 2 above)
Comment (nyh) [reply test/boost/view_schema_pkey_test.cc:723]: Ok. I thought maybe it has some historic significance, but yes, git can fulfill that role.

File: compound_compat.hh (hunk 1 above)
Comment (nyh) [reply compound_compat.hh:74]: @avikivity I looked at the code again, and that "-2" end signal is sort of built into the code... When operator++ finishes iteration, it explicitly sets _offset = -2. It could have been anything else, but it's set to -2. That's why it makes sense for the end iterator to unconditionally return -2, and the begin() of an empty string to match it.
But that being said, 0 is ok too, because if begin==end (and that's the whole point) operator++ is never called so never sets -2 anyway. I'll prepare a patch with the 0.
